src/Components/Web/src/Sections/SectionRegistry.cs

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.AspNetCore.Components.Sections
+{
+    internal class SectionRegistry
+    {
+        private static readonly ConditionalWeakTable<Dispatcher, SectionRegistry> _registries = new();
+
+        private readonly Dictionary<string, List<ISectionContentSubscriber>> _subscribersByName = new();
+        private readonly Dictionary<string, List<ISectionContentProvider>> _providersByName = new();
+
+        public static SectionRegistry GetRegistry(RenderHandle renderHandle)
+            => _registries.GetOrCreateValue(renderHandle.Dispatcher);
+
+        public void AddProvider(string name, ISectionContentProvider provider)
+        {
+            if (!_providersByName.TryGetValue(name, out var providers))
+            {
+                providers = new List<ISectionContentProvider>();
+                _providersByName.Add(name, providers);
+            }
+
+            providers.Add(provider);
+        }
+
+        public void RemoveProvider(string name, ISectionContentProvider provider)
+        {
+            if (!_providersByName.TryGetValue(name, out var providers))
+            {
+                throw new InvalidOperationException($"There are no content providers with the name '{name}'.");
+            }
+
+            var index = providers.LastIndexOf(provider);
+
+            if (index < 0)
+            {
+                throw new InvalidOperationException($"The provider was not found in the providers list of name '{name}'.");
+            }
+
+            providers.RemoveAt(index);
+
+            if (index == providers.Count)
+            {
+                // We just removed the most recently added provider, meaning we need to change
+                // the current content to that of second most recently added provider.
+                var content = GetCurrentProviderContentOrDefault(providers);
+                NotifyContentChangedForAllSubscribers(name, content);
+            }
+        }
+
+        public void Subscribe(string name, ISectionContentSubscriber subscriber)
+        {
+            if (!_subscribersByName.TryGetValue(name, out var subscribers))
```

Comment:
For `<head>` it probably isn't useful, maybe someday if we make this API public then someone may find a use for this. I could change it for this PR if you think it would be worth it.